distributed/shuffle/shuffle_extension.py

```diff
@@ -329,3 +429,151 @@ def _get_shuffle(self, shuffle_id: ShuffleId) -> Shuffle:
             raise ValueError(
                 f"Shuffle {shuffle_id!r} is not registered on worker {self.worker.address}"
             ) from None
+
+    async def close(self):
+        self.executor.shutdown()
+        MultiFile.queue = None
+
+
+class ShuffleSchedulerExtension:
+    """
+    Shuffle extension for the scheduler
+
+    Today this mostly just collects heartbeat messages for the dashboard,
+    but in the future it may be responsible for more
+
+    See Also
+    --------
+    ShuffleWorkerExtension
+    """
+
+    def __init__(self, scheduler):
+        self.scheduler = scheduler
+        self.shuffles = defaultdict(lambda: defaultdict(dict))
+
+    def heartbeat(self, ws, data):
+        for shuffle_id, d in data.items():
+            self.shuffles[shuffle_id][ws.address].update(d)
+
+
+def split_by_worker(
+    df: pd.DataFrame, column: str, npartitions: int, workers: list[str]
+) -> dict:
+    """
+    Split data into many arrow batches, partitioned by destination worker
+    """
+    import numpy as np
+    import pandas as pd
+    import pyarrow as pa
+
+    grouper = (len(workers) * df[column] // npartitions).astype(df[column].dtype).values
+
+    t = pa.Table.from_pandas(df)
+    del df
+    t = t.add_column(len(t.columns), "_worker", [grouper])
+    t = t.sort_by("_worker")
+
+    worker = np.asarray(t.select(["_worker"]))[0]
+    t = t.drop(["_worker"])
+    splits = np.where(worker[1:] != worker[:-1])[0] + 1
+    splits = np.concatenate([[0], splits])
+
+    shards = [
+        t.slice(offset=a, length=b - a) for a, b in toolz.sliding_window(2, splits)
+    ]
+    shards.append(t.slice(offset=splits[-1], length=None))
+
+    w_unique = pd.Series(grouper).unique()
```

Comment:
I think you could use a pyarrow groupby-aggregate for this instead (may not make much difference; can't remember if `Series` creates a copy).

Reply:
Yeah, this is nowhere near a critical point, so I'm leaving this alone for now.  

On the performance track, the next thing to think about is event loop health I think.  But even that is no longer my focus.  Performance is "good enuf" for now.